pbft/pbft_checkpoint_manager.cpp

```diff
@@ -0,0 +1,323 @@
+
+        if (this->latest_local_checkpoint.value().first < cp.first)
+        {
+            this->send_delayed_state_request(cp);
+        }
+        else if (this->latest_local_checkpoint.value().first == cp.first
+                 && this->latest_stable_checkpoint.value().second != cp.second)
+        {
+            LOG(error) << "our checkpoint state disagrees with swarm! sending state request to reconcile";
+            this->send_state_request();
+        }
+    }
+}
+
+void
+pbft_checkpoint_manager::stabilize_checkpoint(const bzn::checkpoint_t& cp)
+{
+    LOG(info) << "Checkpoint at " << cp.first << " is now stable (" << bytes_to_debug_string(cp.second) << ")";
+
+    // update the checkpoint
+    this->latest_stable_checkpoint = cp;
+
+    // clear the old proof
+    for (auto& pair : this->latest_stable_checkpoint_proof)
+    {
+        pair.second.destroy();
+    }
+    this->latest_stable_checkpoint_proof.clear();
+
+    // copy over the new proof
+    for (const auto& pair : this->partial_checkpoint_proofs[cp])
+    {
+        this->latest_stable_checkpoint_proof[pair.first] = {this->storage, pair.second.value(), STABLE_CHECKPOINT_PROOF_KEY, pair.first};
+    }
+
+    // TODO: don't store incomplete proofs seperately
+
+    // clear old data
+
+    // for each partial proof
+    for (auto& map_pair : this->partial_checkpoint_proofs)
+    {
+        const checkpoint_t& this_cp = map_pair.first;
+        // if it's not newer than the newly stabilized proof
+        if (this_cp.first > cp.first)
+        {
+            continue;
+        }
+
+        // delete all the entries
+        for (auto& msg_pair : map_pair.second)
+        {
+            msg_pair.second.destroy();
+        }
+    }
+
+    // now remove them from memory as well
+    for (auto it = this->partial_checkpoint_proofs.begin(); it != this->partial_checkpoint_proofs.end();)
+    {
+        auto seq = it->first.first;
+        if (seq > cp.first)
+        {
+            it++;
+        }
+        else
+        {
+            it = this->partial_checkpoint_proofs.erase(it);
+        }
+    }
+}
+
+void
+pbft_checkpoint_manager::send_delayed_state_request(const checkpoint_t& cp)
+{
+    this->trigger_catchup_timer = this->io_context->make_unique_steady_timer();
+    this->trigger_catchup_timer->expires_from_now(CHECKPOINT_CATCHUP_GRACE_PERIOD);
+    this->trigger_catchup_timer->async_wait(
+            [cp, weak_this = weak_from_this()](auto ec)
+            {
+                if (ec)
+                {
+                    return;
+                }
+
+                auto strong_this = weak_this.lock();
+
+                if (strong_this && strong_this->get_latest_stable_checkpoint() == cp)
+                {
+                    std::lock_guard<std::mutex> lock(strong_this->lock);
+                    strong_this->send_state_request();
+                }
+            }
+    );
+}
+
+void
+pbft_checkpoint_manager::send_state_request()
+{
```

Comment:
KEP-1321 more of the functionality specific to this issue. Sends off a get state request to random (!) peer?

That's cool. Isabel tells me that the request is expensive. And I think that if the random peer is down or something, this procedure will just happen again sometime in the future.